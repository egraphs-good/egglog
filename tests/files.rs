use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
