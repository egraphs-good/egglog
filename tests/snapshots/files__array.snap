---
source: tests/files.rs
expression: serialized
---
nodes:
  String-10923848419941039998:
    op: "\"r1\""
    children: []
    eclass: String-10923848419941039998
    cost: 0
  Var-10923848419941039998:
    op: Var
    children:
      - String-10923848419941039998
    eclass: "0"
    cost: 1
  String-15673665965278112340:
    op: "\"r2\""
    children: []
    eclass: String-15673665965278112340
    cost: 0
  Var-15673665965278112340:
    op: Var
    children:
      - String-15673665965278112340
    eclass: "1"
    cost: 1
  String-1566882640312150612:
    op: "\"r3\""
    children: []
    eclass: String-1566882640312150612
    cost: 0
  Var-1566882640312150612:
    op: Var
    children:
      - String-1566882640312150612
    eclass: "2"
    cost: 1
  select-1351883367118893594:
    op: select
    children:
      - store-8378458532836304618
      - Var-10923848419941039998
    eclass: "6"
    cost: 1
  select-11432643272211809986:
    op: select
    children:
      - store-8378458532836304618
    eclass: "9"
    cost: 1
  select-10654884905055842184:
    op: select
    children:
      - store-9914705668454262402
    eclass: "17"
    cost: 1
  select-11215575473965530962:
    op: select
    children:
      - store-2380122553855171435
    eclass: "17"
    cost: 1
  select-11234845072659508782:
    op: select
    children:
      - mem1-0
    eclass: "9"
    cost: 1
  select-1154085167566592390:
    op: select
    children:
      - AVar-5040379952546458196
    eclass: "17"
    cost: 1
  store-17711188862668393581:
    op: store
    children:
      - store-2380122553855171435
      - Num-11743562013128004906
    eclass: "15"
    cost: 1
  store-8378458532836304618:
    op: store
    children:
      - mem1-0
      - Var-10923848419941039998
      - select-1351883367118893594
    eclass: "5"
    cost: 1
  store-2380122553855171435:
    op: store
    children:
      - AVar-5040379952546458196
      - Num-5871781006564002453
    eclass: "12"
    cost: 1
  store-9914705668454262402:
    op: store
    children:
      - mem1-0
      - Num-11743562013128004906
    eclass: "15"
    cost: 1
  i64-7586556743040283621:
    op: "17"
    children: []
    eclass: i64-7586556743040283621
    cost: 0
  Num-7586556743040283621:
    op: Num
    children:
      - i64-7586556743040283621
    eclass: "7"
    cost: 1
  i64-5871781006564002453:
    op: "1"
    children: []
    eclass: i64-5871781006564002453
    cost: 0
  Num-5871781006564002453:
    op: Num
    children:
      - i64-5871781006564002453
    eclass: "11"
    cost: 1
  i64-11743562013128004906:
    op: "2"
    children: []
    eclass: i64-11743562013128004906
    cost: 0
  Num-11743562013128004906:
    op: Num
    children:
      - i64-11743562013128004906
    eclass: "14"
    cost: 1
  i64-6807129317463932018:
    op: "42"
    children: []
    eclass: i64-6807129317463932018
    cost: 0
  Num-6807129317463932018:
    op: Num
    children:
      - i64-6807129317463932018
    eclass: "6"
    cost: 1
  String-5040379952546458196:
    op: "\"mem1\""
    children: []
    eclass: String-5040379952546458196
    cost: 0
  AVar-5040379952546458196:
    op: AVar
    children:
      - String-5040379952546458196
    eclass: "3"
    cost: 1
  r2-0:
    op: r2
    children: []
    eclass: "1"
    cost: 9223372036854776000
  r3-0:
    op: r3
    children: []
    eclass: "2"
    cost: 9223372036854776000
  mem1-0:
    op: mem1
    children: []
    eclass: "3"
    cost: 9223372036854776000
  test2-0:
    op: test2
    children: []
    eclass: "9"
    cost: 9223372036854776000
  Unit-0:
    op: ()
    children: []
    eclass: Unit-0
    cost: 0
  neq-5871781006564002453:
    op: neq
    children:
      - Var-10923848419941039998
      - r2-0
    eclass: Unit-0
    cost: 1
  neq-7659469028595837896:
    op: neq
    children:
      - Var-15673665965278112340
      - r3-0
    eclass: Unit-0
    cost: 1
  neq-11743562013128004906:
    op: neq
    children:
      - Var-10923848419941039998
      - Var-1566882640312150612
    eclass: Unit-0
    cost: 1
  neq-956286968014291186:
    op: neq
    children:
      - r2-0
      - Var-10923848419941039998
    eclass: Unit-0
    cost: 1
  neq-7784354942592584825:
    op: neq
    children:
      - r3-0
      - Var-15673665965278112340
    eclass: Unit-0
    cost: 1
  neq-1912573936028582372:
    op: neq
    children:
      - Var-1566882640312150612
      - Var-10923848419941039998
    eclass: Unit-0
    cost: 1
  neq-6662973804773207269:
    op: neq
    children:
      - Var-10923848419941039998
    eclass: Unit-0
    cost: 1
  neq-13907576835570464937:
    op: neq
    children:
      - Num-11743562013128004906
      - Num-5871781006564002453
    eclass: Unit-0
    cost: 1
  neq-16618847795869376732:
    op: neq
    children:
      - Num-7586556743040283621
      - Num-5871781006564002453
    eclass: Unit-0
    cost: 1
  neq-1945277965775797206:
    op: neq
    children:
      - Num-5871781006564002453
      - Num-11743562013128004906
    eclass: Unit-0
    cost: 1
root_eclasses: []
class_data:
  "0":
    type: Math
  String-10923848419941039998:
    type: String
  "1":
    type: Math
  String-15673665965278112340:
    type: String
  "2":
    type: Math
  String-1566882640312150612:
    type: String
  "6":
    type: Math
  "5":
    type: Array
  "9":
    type: Math
  "8":
    type: Math
  "17":
    type: Math
  "15":
    type: Array
  "16":
    type: Math
  "12":
    type: Array
  "3":
    type: Array
  "13":
    type: Math
  "14":
    type: Math
  "11":
    type: Math
  "7":
    type: Math
  i64-7586556743040283621:
    type: i64
  i64-5871781006564002453:
    type: i64
  i64-11743562013128004906:
    type: i64
  i64-6807129317463932018:
    type: i64
  String-5040379952546458196:
    type: String
  Unit-0:
    type: Unit

